### DISCUSSION TITLE
Cloudflare Module Workers support / API

### DISCUSSION THREAD

**Discussion by timhanlon** [nitrojs/nitro#679] [Ideas]
Newer Cloudflare Workers features like Durable Objects, service bindings and D1 require the new module syntax.

Building on [this discussion](https://github.com/nuxt/framework/discussions/5487) and [this PR](https://github.com/unjs/nitro/pull/677), I have an [initial proof of concept](https://github.com/timhanlon/nitro/tree/feat/cloudflare-esm) for module Workers. As a bonus, we get access to `waitUntil` in API routes.

Before I submit a PR, I wanted to kick off a discussion about the API. Right now, I'm just stuffing the environment and context from the new module syntax into the existing event context, but this is clearly not ideal:

```
export default defineEventHandler(async event => {
  event.context.ctx.waitUntil(setTimeout(() => console.log('hey'), 5000))

  const res = await event.context.env.WORKER.fetch('https://some.worker.on.workers.dev')
  return await res.json()
})
```

Note: D1 bindings in `wrangler.toml` aren't working yet, `✘ [ERROR] No matching export in ".output/server/index.mjs" for import "default"`

Also, I'm a complete TypeScript rookie, so if anyone can help getting types through to the Nuxt API routes that would be much appreciated.

### Cloudflare docs

- [Migrating to module Workers](https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/)
- [Using Durable Objects](https://developers.cloudflare.com/workers/learning/using-durable-objects/)
- [About Service bindings](https://developers.cloudflare.com/workers/platform/bindings/about-service-bindings/)

**Comment by marshallswain** [comment]
@timhanlon, nice work!  

>  I'm just stuffing the environment and context from the new module syntax into the existing event context, but this is clearly not idea

I actually think what you've done is ideal. If you're concerned it doesn't look pretty, then I agree, but it's not really ugly.  It's just verbose.  The only change I could see as a maybe would be to spread the [`ctx` into `context`](https://github.com/unjs/nitro/compare/main...timhanlon:nitro:feat/cloudflare-esm#diff-af24331bb3186dc52d4f7b371f755ed16ae81079721750bcbbd1b29cd220e696R42), but I don't think it's a good idea from a long-term maintenance perspective.  Leaving `ctx` intact keeps everything that's Cloudflare-API-related bundled up into the same objects that we'd normally use in workers, so it's low mental overhead.
 
In my opinion, this is ready for a PR with a note that says it's built on the PR you referenced.  

TypeScript-wise, the only potential improvement I see would be if we could find types for `cf`, `env`, and `ctx` right here: https://github.com/unjs/nitro/compare/main...timhanlon:nitro:feat/cloudflare-esm#diff-af24331bb3186dc52d4f7b371f755ed16ae81079721750bcbbd1b29cd220e696R43

**Reply by marshallswain** [reply]
I was trying to find the list of everything in Cloudflare's `ctx` object.  It's a small list: https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#syntax-module-worker

**Comment by timhanlon** [comment]
> Note: D1 bindings in wrangler.toml aren't working yet, ✘ [ERROR] No matching export in ".output/server/index.mjs" for import "default"

Just figured out that adding `minify: false` to the preset fixes this.
